examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+                  {
+                    for (unsigned int j=0; j<dofs_per_cell; ++j)
+                      {
+                        local_matrix(i, j) += (  viscosity*scalar_product(grad_phi_u[j], grad_phi_u[i])      //(gradU, gradV_u)
+                                                 + present_velocity_gradients[q]*phi_u[j]*phi_u[i]   //(d_U*grad_U_old, V_u)
+                                                 + grad_phi_u[j]*present_velocity_values[q]*phi_u[i] //(U_old*grad_d_U, V_u)
+                                                 - div_phi_u[i]*phi_p[j]                             //-(d_P, div_V_u)
+                                                 - phi_p[i]*div_phi_u[j]                             //-(div_d_U, V_p)
+                                                 + gamma*div_phi_u[j]*div_phi_u[i]                   //grad-div (div_u_j, div_u_i)
+                                                 + phi_p[i] * phi_p[j]) //pressure mass matrix
+                                              * fe_values.JxW(q);
+
+                      }
+                  }
+
+                if (assemble_rhs)
+                  {
+                    double present_velocity_divergence =  trace(present_velocity_gradients[q]);
+                    local_rhs(i) += (
+                                      -viscosity*scalar_product(present_velocity_gradients[q],grad_phi_u[i]) // -(gradU_old, gradV_u)
+                                      -present_velocity_gradients[q]*present_velocity_values[q]*phi_u[i]     // -(U_old*gradU_old, V_u)
+                                      +present_pressure_values[q]*div_phi_u[i] // +(P_old, div_V_u)
+                                      +present_velocity_divergence*phi_p[i]   // +(div_U_oid, V_p)
+                                      -gamma*present_velocity_divergence*div_phi_u[i]  //grad-div (div_U_old, div_u_i)
+                                    )*fe_values.JxW(q);
+                  }
+
+              }
+          }
+
+        cell->get_dof_indices (local_dof_indices);
+
+        const ConstraintMatrix &constraints_used = initial_step ? nonzero_constraints : zero_constraints;
+
+        // Finally we move pressure mass matrix into a separate matrix:
+
+        if (assemble_matrix)
+          {
+            constraints_used.distribute_local_to_global(local_matrix,
+                                                        local_dof_indices,
+                                                        system_matrix);
+          }
+
+        if (assemble_rhs)
+          {
+            constraints_used.distribute_local_to_global(local_rhs,
+                                                        local_dof_indices,
+                                                        system_rhs);
+          }
+
+      }
+
+    if (assemble_matrix)
+      {
+        pressure_mass_matrix.reinit(sparsity_pattern.block(1,1));
+        pressure_mass_matrix.copy_from(system_matrix.block(1,1));
+        system_matrix.block(1,1) = 0;
```

Comment:
This removes the ones on the diagonal for hanging node pressure dofs from this block. Does this matter?